### PR TITLE
[Conv1d] Check overflow before we compute padding size.

### LINKED ISSUE #161875: Segmentation fault in torch.nn.LazyConv1d
### code
```python
import torch
input_data = torch.randn(1, 3, 32)
lazy_conv1d = torch.nn.LazyConv1d(out_channels=16, kernel_size=3, stride=1,
    padding=9223372036854775803, bias=True)
lazy_conv1d.to(device=torch.device('cpu'))
output = lazy_conv1d(input_data)
```

### output
Segmentation fault

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.17 | packaged by conda-forge | (main, Apr 10 2025, 22:19:12) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.12-2-pve-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A800 80GB PCIe
Nvidia driver version: 550.54.14
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               128
On-line CPU(s) list:                  0-127
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
CPU family:                           6
Model:                                106
Thread(s) per core:                   2
Core(s) per socket:                   32
Socket(s):                            2
Stepping:                             6
CPU(s) scaling MHz:                   100%
CPU max MHz:                          3400.0000
CPU min MHz:                          800.0000
BogoMIPS:                             5200.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            3 MiB (64 instances)
L1i cache:                            2 MiB (64 instances)
L2 cache:                             80 MiB (64 instances)
L3 cache:                             96 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-31,64-95
NUMA node1 CPU(s):                    32-63,96-127
Vulnerability Gather data sampling:   Vulnerable: No microcode
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.6.0
[pip3] torchaudio==2.6.0
[pip3] torchvision==0.21.0
[pip3] triton==3.2.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.6.0                    pypi_0    pypi
[conda] torchaudio                2.6.0                    pypi_0    pypi
[conda] torchvision               0.21.0                   pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi

cc @malfet

### LINKED ISSUE #161877: Aborted in torch.nn.Conv1d
### code
```python
import torch
from torch.nn import Conv1d
input_data = torch.randn(1, 16, 100)


class MyConv1dModule(torch.nn.Module):

    def __init__(self):
        super(MyConv1dModule, self).__init__()
        self.conv1 = Conv1d(in_channels=16, out_channels=32, kernel_size=3,
            stride=1, padding=9223372036854775803)
        self.add_module(name='conv1', module=self.conv1)


model = MyConv1dModule()
output = model.conv1(input_data)
```

### output
realloc(): invalid pointer
realloc(): invalid pointer
realloc(): invalid pointer
Aborted

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.17 | packaged by conda-forge | (main, Apr 10 2025, 22:19:12) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.8.12-2-pve-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A800 80GB PCIe
Nvidia driver version: 550.54.14
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 57 bits virtual
Byte Order:                           Little Endian
CPU(s):                               128
On-line CPU(s) list:                  0-127
Vendor ID:                            GenuineIntel
Model name:                           Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
CPU family:                           6
Model:                                106
Thread(s) per core:                   2
Core(s) per socket:                   32
Socket(s):                            2
Stepping:                             6
CPU(s) scaling MHz:                   100%
CPU max MHz:                          3400.0000
CPU min MHz:                          800.0000
BogoMIPS:                             5200.00
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts vnmi avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                       VT-x
L1d cache:                            3 MiB (64 instances)
L1i cache:                            2 MiB (64 instances)
L2 cache:                             80 MiB (64 instances)
L3 cache:                             96 MiB (2 instances)
NUMA node(s):                         2
NUMA node0 CPU(s):                    0-31,64-95
NUMA node1 CPU(s):                    32-63,96-127
Vulnerability Gather data sampling:   Vulnerable: No microcode
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Vulnerable: Clear CPU buffers attempted, no microcode; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.6.0
[pip3] torchaudio==2.6.0
[pip3] torchvision==0.21.0
[pip3] triton==3.2.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] torch                     2.6.0                    pypi_0    pypi
[conda] torchaudio                2.6.0                    pypi_0    pypi
[conda] torchvision               0.21.0                   pypi_0    pypi
[conda] triton                    3.2.0                    pypi_0    pypi

cc @malfet

### PR DESCRIPTION
**Pull request by thenumberouscode**
Fixes https://github.com/pytorch/pytorch/issues/161877
also fixes https://github.com/pytorch/pytorch/issues/161875

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168

### CODE REVIEW

**Comment by arkadip-maitra** [comment]
Add test

**Comment by arkadip-maitra** [comment]
@thenumberouscode it works but idk if this is where the check should be done. also check once if the problem is there when you pass transposed as False since you are checking inside the if condition `if (!transposed)`. also pass device argument in the test function so cuda gets tested too.

**Comment by thenumberouscode** [comment]
> @thenumberouscode it works but idk if this is where the check should be done. also check once if the problem is there when you pass transposed as False since you are checking inside the if condition `if (!transposed)`. also pass device argument in the test function so cuda gets tested too.

@arkadip-maitra Thanks for reviewing, your suggestion is indeed very useful. I've add transposed padding size check, and the unit test will automatically run on different device based on my understanding. correct me if I'm wrong.

**Comment by thenumberouscode** [comment]
@Skylion007 Could you help me to run the workflow again? thanks

**Comment by thenumberouscode** [comment]
@bdhirsh Could you please help me to run the workflow again? thanks

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -688,6 +688,11 @@ static void check_shape_forward(const at::Tensor& input,
              ", but got bias of size ", at::symint::sizes<T>(bias), " instead");
 
     for (const auto i : c10::irange(2, k)) {
+      // std::numeric_limits<T>::max can be 0 when T is SymInt
+      // So hardcode T to int64_t to get correct value of padding limit size.
+      TORCH_CHECK(padding[i-2] <= (std::numeric_limits<int64_t>::max() - padding[i-2]),
+                  "Given padding=", padding[i-2], " at dimension ", i-2, " , expected padding to be at most ",
+                  (std::numeric_limits<int64_t>::max() / 2));
```
Comment (jbschlosser) [inline aten/src/ATen/native/Convolution.cpp]: correct me if I'm wrong, but I don't think this check properly guards against a large enough padding value when `T` is `int32_t`. I expect the check to pass perfectly fine and we still get overflow on the next line when calculating `2 * padding[i-2]`.

Comment (thenumberouscode) [reply aten/src/ATen/native/Convolution.cpp]: correct me if I'm wrong, After reviewing the relevant code, T can only be one of two types: int64_t or SymInt. SymInt utilizes int64_t as its underlying data structure for value computation.

code snippet:
SymInt's data_ member is an int64_t, which is used in its operator computations.
https://github.com/pytorch/pytorch/blob/b11593c31bd84845e1573de0c15692387c572a2f/c10/core/SymInt.h#L431-L432

The comments indicate that ConvParam uses int64_t in eager mode.
https://github.com/pytorch/pytorch/blob/b11593c31bd84845e1573de0c15692387c572a2f/aten/src/ATen/native/Convolution.cpp#L283-L288

Calls from Python code will consistently use ConvParam<c10::SymInt>.
https://github.com/pytorch/pytorch/blob/b11593c31bd84845e1573de0c15692387c572a2f/aten/src/ATen/native/Convolution.cpp#L1301-L1312

Comment (jbschlosser) [reply aten/src/ATen/native/Convolution.cpp]: @thenumberouscode fair enough re: the callsites, I was just reviewing this in isolation. I do think it's a good property for the code in this helper to be correct for all possible T values even if today only a couple are exercised.

wdyt about having this helper utilize a new templated private helper for getting the max value for a type T? it'd return `std::numeric_limits<T>::max()` for all types with a template specialization for `c10::SymInt` that returns `std::numeric_limits<int64_t>::max()`. That way, this helper is a bit more future-proof.

Comment (thenumberouscode) [reply aten/src/ATen/native/Convolution.cpp]: Very good suggestion. please review the code again to see if any changes is still needed。

File: c10/core/SymInt.h [hunk 2]
```diff
@@ -556,3 +556,26 @@ inline SymBool sym_ge(const SymInt& a, const SymInt& b) {
...
+
+  static constexpr int64_t min() noexcept {
+    return std::numeric_limits<int64_t>::min();
+  }
+
+  static constexpr bool is_signed = true;
+  static constexpr bool is_integer = true;
+};
+
+} // namespace std
```
Comment (jbschlosser) [inline c10/core/SymInt.h:581]: cc @ezyang to verify there's no unintended downsides to specializing `numeric_limits` for `c10::SymInt`? if this is too risky, I'll suggest some private templatized helper for getting the max in the conv check logic instead of using `numeric_limits` itself

Comment (thenumberouscode) [reply c10/core/SymInt.h:581]: @ezyang Could you please take a look?

Comment (jbschlosser) [reply c10/core/SymInt.h:581]: @thenumberouscode FYI the PyTorch Conference is going on at the moment so there may be some delay :)

Comment (ezyang) [reply c10/core/SymInt.h:581]: Endorsing this change

**Review by jbschlosser — APPROVED** [review]
Approving, thanks! just need to fix lint so we can merge
